File: Analizador Semantico/Analizador/src/Arbol.java
import java.util.ArrayList;
import java.util.List;

public class Arbol {
    private Nodo raiz;
    private Postfix poshelp;

    public Arbol(Nodo raiz) {
        this.raiz = raiz;
        this.poshelp = new Postfix(new ArrayList<>());
    }

    public void recorrer() {
        for (int index = 0; index < this.raiz.hijos.size(); index++) {
            Nodo n = this.raiz.hijos.get(index);
            TokenType t = n.value.type;
            switch (t) {
                case ADD, SUB, MULT, DIAG -> {
                    SolverAritmetico solver = new SolverAritmetico();
                    Object res = solver.resolver(n);
                    System.out.println(res);
                }
                case VAR -> solverVar(n);
                case IF -> resolverIf(n);
                case WHILE -> solverWhile(n);
                case FOR -> solverFor(n);
                case GREAT_EQUAL, EQUAL, GREAT, LESS_EQUAL, LESS_THAN, DIFERENT -> {
                    SolverRel solver = new SolverRel();
                    Object res = solver.resolver(n);
                    System.out.println(res);
                }
                case AND, OR -> {
                    SolverLogic solver = new SolverLogic();
                    Object res = solver.resolver(n);
                    System.out.println(res);
                }
                case PRINT -> {
                    Object res = solverPrint(n);
                    System.out.println(res);
                }
                case ASIGNATION -> solverAsig(n);
            }
        }
    }

    public void resolverIf(Nodo n) {
        if (n.hijos == null) {
            if (n.value.type == TokenType.NUMBER || n.value.type == TokenType.STRING) {
                return n.value.literal;
            } else if (n.value.type == TokenType.IDENTIFIER) {
                return ts.simbolos.obtener(n.value.lexeme);
            }
        }
        Nodo cond = n.hijos.get(0);
        boolean rcond = checkCond(cond);
        if (n.hijos.get(n.hijos.size() - 1).value.type == TokenType.ELSE) {
            boolean el = true;
            List<Nodo> body = n.hijos.subList(1, n.hijos.size() - 1);
            List<Nodo> ebody = n.hijos.get(n.hijos.size() - 1).hijos;
        } else {
            List<Nodo> body = n.hijos.subList(1, n.hijos.size());
            boolean el = false;
        }
        if (rcond) {
            Nodo raiz = new Nodo(new Token(TokenType.NULL, "", "", null));
            raiz.insertar_hijos(body);
            Arbol araux = new Arbol(raiz);
            araux.recorrer();
        } else {
            if (el) {
                Nodo raiz = new Nodo(new Token(TokenType.NULL, "", "", null));
                raiz.insertar_hijos(ebody);
                Arbol araux = new Arbol(raiz);
                araux.recorrer();
            } else {
                return;
            }
        }
    }

    public void solverWhile(Nodo n) {
        Nodo cond = n.hijos.get(0);
        List<Nodo> body = n.hijos.subList(1, n.hijos.size());
        while (checkCond(cond)) {
            Nodo raiz = new Nodo(new Token(TokenType.NULL, "", "", null));
            raiz.insertar_hijos(body);
            Arbol araux = new Arbol(raiz);
            araux.recorrer();
        }
    }

    public boolean checkCond(Nodo cond) {
        TokenType type = cond.value.type;
        switch (type) {
            case GREAT_EQUAL, EQUAL, GREAT, LESS_EQUAL, LESS_THAN, DIFERENT -> {
                SolverRel solver = new SolverRel();
                boolean rcond = solver.resolver(cond);
                return rcond;
            }
            case AND, OR -> {
                SolverLogic solver = new SolverLogic();
                boolean rcond = solver.resolver(cond);
                return rcond;
            }
            case TRUE -> {
                boolean rcond = true;
                return rcond;
            }
            case FALSE -> {
                boolean rcond = false;
                return rcond;
            }
            case IDENTIFIER -> {
                if (ts.simbolos.existeIdentificador(cond.value.lexeme)) {
                    if (ts.simbolos.obtener(cond.value.lexeme) instanceof Boolean) {
                        boolean rcond = (boolean) ts.simbolos.obtener(cond.value.lexeme);
                        return rcond;
                    } else {
                        System.out.println("Error la variable evaluada no es un boleano.\n");
                        System.exit(0);
                    }
                } else {
                    System.out.println("Error: La variable " + cond.value.lexeme + " no existe.\n");
                    System.exit(0);
                }
            }
            default -> {
                System.out.println("Error: El resultado debe ser un boleano");
                System.exit(0);
            }
        }
    }

    public void solverFor(Nodo n) {
        Nodo ini = n.hijos.get(0);
        Nodo cond = n.hijos.get(1);
        Nodo increase = n.hijos.get(2);
        List<Nodo> body = n.hijos.subList(3, n.hijos.size());
        solverVar(n.hijos.get(0));
        while (checkCond(cond)) {
            Nodo raiz = new Nodo(new Token(TokenType.NULL, "", "", null));
            raiz.insertar_hijos(body);
            Arbol araux = new Arbol(raiz);
            araux.recorrer();
            solverAsig(increase);
        }
    }

    public void solverVar(Nodo n) {
        if (n.hijos.size() == 1) {
            if (ts.simbolos.existeIdentificador(n.hijos.get(0).value.lexeme)) {
                System.out.println("Error: La variable " + n.hijos.get(0).value.lexeme + " ya existe");
                return;
            }
            ts.simbolos.asignar(n.hijos.get(0).value.lexeme, null);
            return;
        } else if (n.hijos.size() == 2) {
            if (ts.simbolos.existeIdentificador(n.hijos.get(0).value.lexeme)) {
                System.out.println("Error: La variable " + n.hijos.get(0).value.lexeme + " ya existe");
                return;
            } else {
                String key = n.hijos.get(0).value.lexeme;
            }
            if (poshelp.esOperador(n.hijos.get(1).value.type)) {
                TokenType type = n.hijos.get(1).value.type;
                Object value;
                switch (type) {
                    case ADD, SUB, MULT, DIAG -> {
                        SolverAritmetico solver = new SolverAritmetico();
                        value = solver.resolver(n.hijos.get(1));
                    }
                    case GREAT_EQUAL, EQUAL, GREAT, LESS_EQUAL, LESS_THAN, DIFERENT -> {
                        SolverRel solver = new SolverRel();
                        value = solver.resolver(n.hijos.get(1));
                    }
                    case AND, OR -> {
                        SolverLogic solver = new SolverLogic();
                        value = solver.resolver(n.hijos.get(1));
                    }
                }
            } else if (n.hijos.get(1).value.type == TokenType.IDENTIFIER) {
                if (ts.simbolos.existeIdentificador(n.hijos.get(1).value.lexeme)) {
                    Object value = ts.simbolos.obtener(n.hijos.get(1).value.lexeme);
                    return;
                } else {
                    System.out.println("Error: La variable " + n.hijos.get(1).value.lexeme + " no existe, por lo cual no puede ser asignada a " + n.hijos.get(0).value.lexeme + "\n");
                    System.exit(0);
                }
            } else {
                Object value = n.hijos.get(1).value.literal;
            }
            ts.simbolos.asignar(key, value);
            return;
        } else {
            System.out.println("Error al declarar la variable");
            return;
        }
    }

    public void solverAsig(Nodo n) {
        if (ts.simbolos.existeIdentificador(n.hijos.get(0).value.lexeme)) {
            if (n.hijos.get(0).value.type == TokenType.IDENTIFIER) {
                if (poshelp.esOperador(n.hijos.get(1).value.type)) {
                    TokenType type = n.hijos.get(1).value.type;
                    Object value;
                    switch (type) {
                        case ADD, SUB, MULT, DIAG -> {
                            SolverAritmetico solver = new SolverAritmetico();
                            value = solver.resolver(n.hijos.get(1));
                        }
                        case AND, OR -> {
                            SolverLogic solver = new SolverLogic();
                            value = solver.resolver(n.hijos.get(1));
                        }
                        case GREAT_EQUAL, EQUAL, GREAT, LESS_EQUAL, LESS_THAN, DIFERENT -> {
                            SolverRel solver = new SolverRel();
                            value = solver.resolver(n.hijos.get(1));
                        }
                    }
                    ts.simbolos.reasig(n.hijos.get(0).value.lexeme, value);
                    return;
                } else {
                    ts.simbolos.reasig(n.hijos.get(0).value.lexeme, n.hijos.get(1).value.literal);
                    return;
                }
            }
        } else {
            System.out.println("Error: La variable " + n.hijos.get(0).value.lexeme + " no existe");
            return;
        }
    }

    public Object solverPrint(Nodo n) {
        if (n.hijos == null) {
            if (n.value.type == TokenType.NUMBER || n.value.type == TokenType.STRING) {
                return n.value.literal;
            } else if (n.value.type == TokenType.IDENTIFIER) {
                return ts.simbolos.obtener(n.value.lexeme);
            }
        }
        Nodo hijo = n.hijos.get(0);
        if (poshelp.esOperador(hijo.value.type)) {
            TokenType type = hijo.value.type;
            Object res;
            switch (type) {
                case ADD, SUB, MULT, DIAG -> {
                    SolverAritmetico solver = new SolverAritmetico();
                    res = solver.resolver(hijo);
                    return res;
                }
                case AND, OR -> {
                    SolverLogic solver = new SolverLogic();
                    res = solver.resolver(hijo);
                    return res;
                }
                case GREAT_EQUAL, EQUAL, GREAT, LESS_EQUAL, LESS_THAN, DIFERENT -> {
                    SolverRel solver = new SolverRel();
                    res = solver.resolver(hijo);
                    return res;
                }
            }
        } else {
            Object valor = solverPrint(hijo);
            return valor;
        }
    }
}

File: AnalizadorSem/Arbol.py
from Tokentype import TokenType
from Token import Token
from Nodo import Nodo
from Aritmetico import SolverAritmetico
import SymbolTable as ts
from Relacion import SolverRel
from Logic import SolverLogic
from Postfija import Postfix
import sys


class Arbol:
    def __init__(self, raiz: Nodo) -> None:
        self.raiz = raiz
        self.poshelp = Postfix([])

    def recorrer(self):

        for index, n in enumerate(self.raiz.hijos):
            t = n.value
            match t.type:
                case TokenType.ADD | TokenType.SUB | TokenType.MULT | TokenType.DIAG:
                    solver = SolverAritmetico()
                    res = solver.resolver(n)
                    print(f"{res}")
                case TokenType.VAR:
                    self.solverVar(n)

                case TokenType.IF:
                    self.resolverIf(n)
                case TokenType.WHILE:
                    self.solverWhile(n)
                case TokenType.FOR:
                    self.solverFor(n)
                    pass
                case TokenType.GREAT_EQUAL | TokenType.EQUAL | TokenType.GREAT | TokenType.LESS_EQUAL | TokenType.LESS_THAN | TokenType.DIFERENT:
                    solver = SolverRel()
                    res = solver.resolver(n)
                    print(f"{res}")
                case TokenType.AND | TokenType.OR:
                    
                    solver = SolverLogic()
                    res = solver.resolver(n)
                    print(f"{res}")
                case TokenType.PRINT:
                    res = self.solverPrint(n)
                    print(f"{res}")
                    pass
                case TokenType.ASIGNATION:
                    self.solverAsig(n)
                    pass

    def resolverIf(self, n: Nodo):
        if n.hijos is None:
            if n.value.type == TokenType.NUMBER or n.value.type == TokenType.STRING:
                return n.value.literal
            elif n.value.type == TokenType.IDENTIFIER:
                return ts.simbolos.obtener(n.value.lexeme)

        cond = n.hijos[0]
        rcond = self.checkCond(cond)

        if n.hijos[-1].value.type == TokenType.ELSE:
            el = True
            body = n.hijos[1:]
            ebody = n.hijos[-1].hijos
        else:
            body = n.hijos[1:]
            el = False

        if rcond:
            raiz = Nodo(Token(TokenType.NULL, "", "", None))
            raiz.insertar_hijos(body)
            araux = Arbol(raiz)
            araux.recorrer()
        else:
            if el:
                raiz = Nodo(Token(TokenType.NULL, "", "", None))
                raiz.insertar_hijos(ebody)
                araux = Arbol(raiz)
                araux.recorrer()
            else:
                pass

    def solverWhile(self, n: Nodo):

        cond = n.hijos[0]
        body = n.hijos[1:]

        while self.checkCond(cond):
            raiz = Nodo(Token(TokenType.NULL, "", "", None))
            raiz.insertar_hijos(body)
            araux = Arbol(raiz)
            araux.recorrer()

    def checkCond(self, cond: Nodo) -> bool:
        match cond.value.type:
            case TokenType.GREAT_EQUAL | TokenType.EQUAL | TokenType.GREAT | TokenType.LESS_EQUAL | TokenType.LESS_THAN | TokenType.DIFERENT:
                solver = SolverRel()
                rcond = solver.resolver(cond)
                return rcond
            case TokenType.AND | TokenType.OR:

                solver = SolverLogic()
                rcond = solver.resolver(cond)
                return rcond
            case TokenType.TRUE:
                rcond = True
                return rcond
            case TokenType.FALSE:
                rcond = False
                return rcond
            case TokenType.IDENTIFIER:
                if ts.simbolos.existeIdentificador(cond.value.lexeme):
                    if isinstance(ts.simbolos.obtener(cond.value.lexeme), bool):
                        rcond = ts.simbolos.obtener(cond.value.lexeme)
                        return rcond
                    else:
                        print("Error la variable evaluada no es un boleano.\n")
                        sys.exit()
                else:
                    print(f"Error: La variable {cond.value.lexeme} no existe.\n")
                    sys.exit()
            case _:
                print("Error: El resultado debe ser un boleano")
                sys.exit()

    def solverFor(self, n: Nodo):
        ini = n.hijos[0]
        cond = n.hijos[1]
        increase = n.hijos[2]
        body = n.hijos[3:]

        self.solverVar(n.hijos[0])

        while self.checkCond(cond):
            raiz = Nodo(Token(TokenType.NULL, "", "", None))
            raiz.insertar_hijos(body)
            araux = Arbol(raiz)
            araux.recorrer()
            self.solverAsig(increase)

        pass

    def solverVar(self, n: Nodo):

        if len(n.hijos) == 1:
            if ts.simbolos.existeIdentificador(n.hijos[0].value.lexeme):
                print(f"Error: La variable {n.hijos[0].value.lexeme} ya existe")
                return
            ts.simbolos.asignar(n.hijos[0].value.lexeme, None)
            return
        elif len(n.hijos) == 2:

            if ts.simbolos.existeIdentificador(n.hijos[0].value.lexeme):
                print(f"Error: La variable {n.hijos[0].value.lexeme} ya existe")
                return
            else:
                key = n.hijos[0].value.lexeme

            if self.poshelp.esOperador(n.hijos[1].value.type):
                match n.hijos[1].value.type:
                    case TokenType.ADD | TokenType.SUB | TokenType.MULT | TokenType.DIAG:
                        solver = SolverAritmetico()
                        value = solver.resolver(n.hijos[1])
                    case TokenType.GREAT_EQUAL | TokenType.EQUAL | TokenType.GREAT | TokenType.LESS_EQUAL | TokenType.LESS_THAN | TokenType.DIFERENT:
                        solver = SolverRel()
                        value = solver.resolver(n.hijos[1])
                    case TokenType.AND | TokenType.OR:
                        solver = SolverLogic()
                        value = solver.resolver(n.hijos[1])



            elif n.hijos[1].value.type == TokenType.IDENTIFIER:
                if ts.simbolos.existeIdentificador(n.hijos[1].value.lexeme):
                    value = ts.simbolos.obtener(n.hijos[1].value.lexeme)
                    pass
                else:
                    print(
                        f"Error: La variable {n.hijos[1].value.lexeme} no existe, por lo cual no puede ser asignada a {n.hijos[0].value.lexeme}\n")
                    sys.exit()

            else:
                value = n.hijos[1].value.literal
            ts.simbolos.asignar(key, value)
            return
        else:
            print("Error al declarar la variable")
            return None

    def solverAsig(self, n: Nodo):

        if ts.simbolos.existeIdentificador(n.hijos[0].value.lexeme):

            if n.hijos[0].value.type == TokenType.IDENTIFIER:
                if self.poshelp.esOperador(n.hijos[1].value.type):

                    match n.hijos[1].value.type:
                        case TokenType.ADD | TokenType.SUB | TokenType.MULT | TokenType.DIAG:
                            solver = SolverAritmetico()
                            value = solver.resolver(n.hijos[1])
                        case TokenType.GREAT_EQUAL | TokenType.EQUAL | TokenType.GREAT | TokenType.LESS_EQUAL | TokenType.LESS_THAN | TokenType.DIFERENT:
                            solver = SolverRel()
                            value = solver.resolver(n.hijos[1])
                        case TokenType.AND | TokenType.OR:
                            solver = SolverLogic()
                            value = solver.resolver(n.hijos[1])

                    ts.simbolos.reasig(n.hijos[0].value.lexeme, value)
                    return
                else:
                    ts.simbolos.reasig(n.hijos[0].value.lexeme, n.hijos[1].value.literal)
                    return
        else:
            print(f"Error: La variable {n.hijos[0].value.lexeme} no existe")
            return

    def solverPrint(self, n: Nodo):

        if n.hijos is None:
            if n.value.type == TokenType.NUMBER or n.value.type == TokenType.STRING:
                return n.value.literal
            elif n.value.type == TokenType.IDENTIFIER:
                return ts.simbolos.obtener(n.value.lexeme)

        hijo: Nodo = n.hijos[0]

        if self.poshelp.esOperador(hijo.value.type):
            match hijo.value.type:
                case TokenType.ADD | TokenType.SUB | TokenType.MULT | TokenType.DIAG:
                    solver = SolverAritmetico()
                    res = solver.resolver(hijo)
                    return res
                case TokenType.AND | TokenType.OR:

                    solver = SolverLogic()
                    res = solver.resolver(hijo)
                    return res
                case TokenType.GREAT_EQUAL | TokenType.EQUAL | TokenType.GREAT | TokenType.LESS_EQUAL | TokenType.LESS_THAN | TokenType.DIFERENT:
                    solver = SolverRel()
                    res = solver.resolver(hijo)
                    return res
        else:
            valor = self.solverPrint(hijo)
            return valor
